layouts/fastly/helix.vcl

```diff
@@ -10,6 +10,130 @@
 # governing permissions and limitations under the License.
 #
 
+
+# About URLs and Host headers:
+# This VCL will use the following conventions:
+# `req.url` - The canonical URL, will be used to determine the hash of the object
+# `req.http.X-Orig-URL` - The URL as received from the client
+# `req.http.X-Backend-URL` - The URL as should be sent to the origin (backend)
+#
+# Having `req.url` be canonical, vs `X-Backend-URL` being what's sent to origin 
+# allows us to do restarts to find content on other origins, and use origin
+# specific URLs, yet have the object stored under a single hash, meaning that 
+# a cache lookup will find the proper object on the first attempt once an 
+# object is in the cache.
+#
+# So `req.url` is what should be left after filtering query string parameters,
+# doing normalizations, etc. Then `req.http.X-Backend-URL` is a transformation of
+# that, using the origin specific rewriting, etc.
+#
+# Domain names are handled a bit simpler:
+# `req.http.Host` - The canonical domain name for the site, used for the hash
+# `req.http.X-Orig-Host` - The original domain name from the client
+#
+# `bereq.http.Host` will be set based on the backend used.
+#
+# When forwarding the request from edge POP to shield POP, the original URL and
+# original domain name will be used.
+
+
+/**
+ * Check and set the `X-From-Edge` header, which allows us to determine whether
+ * a request is coming from a Fastly edge POP. Use this instead of `Fastly-FF`
+ * since that can be spoofed, or come from another Fastly service.
+ */
+sub hlx_check_from_edge {
+  # Verify correct format
+  if (req.http.X-From-Edge ~ "^(([0-9]+),[^,]+),(0x[0-9a-f]{64})$"
+      # Verify the HMAC
+      && digest.secure_is_equal(
+           # Using existing secret, but adding it to service id
+           # TODO: use a separate secret, because changing it can have some
+           # gnarly effects, if you don't follow it with a purge all
+           digest.hmac_sha256(req.service_id + table.lookup(secrets, "OPENWHISK_AUTH")
+           , re.group.1), re.group.3)) {
+    declare local var.time TIME;
+    set var.time = std.time(re.group.2, std.integer2time(-1));
+    # And verify the timestamp is not off by more than 2 seconds
+    if (time.is_after(var.time, time.sub(now, 2s))
+        && time.is_after(time.add(now, 2s), var.time)) {
+      # Do nothing, the header is valid
+      return;
+    }
+  }
+  # Header isn't valid, chuck it
+  unset req.http.X-From-Edge;
+}
+
+sub hlx_set_from_edge {
```

Comment:
As someone looking at this VCL a fresh, personally I'd find it helpful if these subs had a bit better annotation and some context. it would help me better understand the intent if had a description and maybe even from where in the state machine it's intended to be called. Quite subjective though, so feel free to ignore.